File: python/two_player_unicycle_4d.py
"""
BSD 3-Clause License

Copyright (c) 2019, HJ Reachability Group
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of the copyright holder nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Author(s): David Fridovich-Keil ( dfk@eecs.berkeley.edu )
"""
################################################################################
#
# 4D unicycle model with disturbance. Dynamics are as follows:
#                          \dot x     = v cos theta + u21
#                          \dot y     = v sin theta + u22
#                          \dot theta = u11
#                          \dot v     = u12
#
################################################################################

import torch
import numpy as np

from multiplayer_dynamical_system import MultiPlayerDynamicalSystem

class TwoPlayerUnicycle4D(MultiPlayerDynamicalSystem):
    """ 4D unicycle model with disturbances. """

    def __init__(self, T=0.1):
        super(TwoPlayerUnicycle4D, self).__init__(4, [2, 2], T)

    def __call__(self, x, u):
        """
        Compute the time derivative of state for a particular state/control.
        NOTE: `x`, and all `u` should be 2D (i.e. column vectors).

        :param x: current state
        :type x: torch.Tensor or np.array
        :param u: list of current control inputs for all each player
        :type u: [torch.Tensor] or [np.array]
        :return: current time derivative of state
        :rtype: torch.Tensor or np.array
        """
        assert len(u) == self._num_players

        if isinstance(x, np.ndarray):
            x_dot = np.zeros((self._x_dim, 1))
            cos = np.cos
            sin = np.sin
        else:
            x_dot = torch.zeros((self._x_dim, 1))
            cos = torch.cos
            sin = torch.sin

        x_dot[0, 0] = x[3, 0] * cos(x[2, 0]) + u[1][0, 0]
        x_dot[1, 0] = x[3, 0] * sin(x[2, 0]) + u[1][1, 0]
        x_dot[2, 0] = u[0][0, 0]
        x_dot[3, 0] = u[0][1, 0]
        return x_dot

File: src/two_player_unicycle_4d.cpp
/*
 * Copyright (c) 2019, The Regents of the University of California (Regents).
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *    1. Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 *    3. Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS AS IS
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Please contact the author(s) of this library if you have any questions.
 * Authors: David Fridovich-Keil   ( dfk@eecs.berkeley.edu )
 */

///////////////////////////////////////////////////////////////////////////////
//
// Two player dynamics modeling a unicycle with velocity disturbance.
// State is [x, y, theta, v], u1 is [omega, a], u2 is [dx, dy] and dynamics are:
//                     \dot px    = v cos theta + dx
//                     \dot py    = v sin theta + dy
//                     \dot theta = omega
//                     \dot v     = a
//
///////////////////////////////////////////////////////////////////////////////

#include <ilqgames/dynamics/two_player_unicycle_4d.h>

namespace ilqgames {

// Constexprs for state indices.
const Dimension TwoPlayerUnicycle4D::kNumXDims = 4;
const Dimension TwoPlayerUnicycle4D::kPxIdx = 0;
const Dimension TwoPlayerUnicycle4D::kPyIdx = 1;
const Dimension TwoPlayerUnicycle4D::kThetaIdx = 2;
const Dimension TwoPlayerUnicycle4D::kVIdx = 3;

// Constexprs for control indices.
const PlayerIndex TwoPlayerUnicycle4D::kNumPlayers = 2;

const Dimension TwoPlayerUnicycle4D::kNumU1Dims = 2;
const Dimension TwoPlayerUnicycle4D::kOmegaIdx = 0;
const Dimension TwoPlayerUnicycle4D::kAIdx = 1;

const Dimension TwoPlayerUnicycle4D::kNumU2Dims = 2;
const Dimension TwoPlayerUnicycle4D::kDxIdx = 0;
const Dimension TwoPlayerUnicycle4D::kDyIdx = 1;

}